org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/generators/AbstractXML2GrammarGenerator.java

```diff
@@ -0,0 +1,164 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.extensions.generators;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.lemminx.commons.TextDocument;
+import org.eclipse.lemminx.dom.DOMAttr;
+import org.eclipse.lemminx.services.IXMLFullFormatter;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lemminx.utils.StringUtils;
+import org.eclipse.lemminx.utils.XMLBuilder;
+import org.w3c.dom.Attr;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+/**
+ * Abstract class to generate a grammar (XSD, DTD, etcc) from a given XML
+ * source.
+ * 
+ * <p>
+ * The process is to build a generic {@link Grammar} instance from the XML
+ * source and each implementation uses this grammar information to generate the
+ * proper grammar.
+ * </p>
+ * 
+ * @param <T> the grammar generator settings.
+ */
+public abstract class AbstractXML2GrammarGenerator<T extends FileContentGeneratorSettings>
+		implements IFileContentGenerator<Document, T> {
+
+	@Override
+	public String generate(Document document, SharedSettings sharedSettings, T generatorSettings,

```

Comment:
Why you want I change that? @xorye @fbricon what do you think about this rename?

Reply:
Sorry, I should have mentioned that. I think it will help to make it more clear that the document being passed in is the one to generate off of. I think with the current name it could be confused for a blank document that gets passed in, then is populated by the generator.

Reply:
fixed, I rename to `sourceDocument`

Reply:
I see what you mean, but I'm leaning towards `document` since its brief and the javadoc for `IFileContentGenerator#generate` explains what each parameter is